File: bindings/python/proven/safe_calculator.py
# SPDX-License-Identifier: PMPL-1.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

"""
SafeCalculator - Expression evaluator with overflow and division protection.

Evaluates mathematical expressions safely without crashes or security issues.
All expression evaluation is delegated to the Idris core via FFI.
"""

from typing import Optional, Dict

from .core import ProvenStatus, get_lib


class SafeCalculator:
    """Safe mathematical expression evaluator via FFI."""

    @staticmethod
    def evaluate(expression: str, variables: Optional[Dict[str, float]] = None) -> Optional[float]:
        """
        Evaluate a mathematical expression safely via FFI.

        Args:
            expression: Mathematical expression string
            variables: Optional variable bindings

        Returns:
            Result, or None if expression is invalid or would overflow

        Example:
            >>> SafeCalculator.evaluate("2 + 3 * 4")
            14.0
            >>> SafeCalculator.evaluate("sqrt(16) + x", {"x": 5})
            9.0
            >>> SafeCalculator.evaluate("1 / 0")
            None
        """
        if not expression:
            return None

        # Substitute variables into the expression string
        expr = expression
        if variables:
            for name, value in variables.items():
                expr = expr.replace(name, str(value))

        lib = get_lib()
        encoded = expr.encode("utf-8")
        result = lib.proven_calculator_evaluate(encoded, len(encoded))
        if result.status != ProvenStatus.OK:
            return None
        return result.value

File: bindings/haskell/src/Proven/SafeCalculator.hs
{- SPDX-License-Identifier: PMPL-1.0-or-later -}
{- Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk> -}

-- | Safe expression evaluation via libproven FFI.
--
-- Expression parsing and evaluation are performed by the
-- Idris 2 verified core.
module Proven.SafeCalculator
  ( evalExpression
  ) where

import Proven.FFI (c_proven_calculator_eval)
import Proven.Core (withCStringLen', floatResultToMaybe)

-- | Evaluate a mathematical expression string.
-- Delegates to @proven_calculator_eval@ in libproven.
evalExpression :: String -> IO (Maybe Double)
evalExpression expr = withCStringLen' expr $ \ptr len ->
  floatResultToMaybe <$> c_proven_calculator_eval ptr len
